EIPS/eip-3978.md

```diff
@@ -0,0 +1,49 @@
+---
+eip: 3978
+title: Gas refunds on reverts
+description: Do not erase gas refunds on transaction subcall reverts, due users pay a lot of gas for storage non-modification.
+author: Anton Bukov (@k06a), Mikhail Melnik (@ZumZoom)
+discussions-to: https://ethereum-magicians.org/t/eip-3978-gas-refunds-on-reverts/7071/
+status: Draft
+type: Standards Track
+category: Core
+created: 2021-09-16
+---
+
+## Abstract
+
+Since [EIP-3298](https://eips.ethereum.org/EIPS/eip-3298) gas refunds works for storage restores only inside the same transaction. For example [ERC-20](./eip-20.md) `approve` + `transferFrom` flow between 2 smart contracts according to [EIP-2200](https://eips.ethereum.org/EIPS/eip-2200) and [EIP-2929](https://eips.ethereum.org/EIPS/eip-2929) will cost nearly to `21600` gas with gas refund counter `20000`. But in case of reverting this subcall (containing both `approve` and `transferForm`) gas refund will be erased, while smart contract storage will remain unmodified. I think it should keep storage access costs, but still refund modification costs.
```

Comment:
```suggestion
 Since [EIP-3298](./eip-3298.md) gas refunds works for storage restores only inside the same transaction. For example [ERC-20](./eip-20.md) `approve` + `transferFrom` flow between 2 smart contracts according to [EIP-2200](./eip-2200.md) and [EIP-2929](./eip-2929.md) will cost nearly to `21600` gas with gas refund counter `20000`. But in case of reverting this subcall (containing both `approve` and `transferForm`) gas refund will be erased, while smart contract storage will remain unmodified. I think it should keep storage access costs, but still refund modification costs.
```
Note that the reference to EIP-3298 will cause this EIP to depend on 3298.  Consider removing that reference if you don't want this EIP to be held up by the progress of that EIP.